kiwi/utils/rpm_database.py

```diff
@@ -0,0 +1,139 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with kiwi.  If not, see <http://www.gnu.org/licenses/>
+#
+import os
+
+# project
+from kiwi.command import Command
+from kiwi.path import Path
+from kiwi.utils.rpm import Rpm
+from kiwi.exceptions import KiwiRpmDatabaseReloadError
+
+
+class RpmDataBase(object):
+    """
+    **Setup RPM database configuration**
+    """
+    def __init__(self, root_dir):
+        self.rpmdb_host = Rpm()
+        self.rpmdb_image = Rpm(root_dir)
+        self.root_dir = root_dir
+
+    def has_rpm(self):
+        """
+        Check if rpm binary was found in root_dir
+        """
+        rpm_search_env = {
+            'PATH': os.sep.join([self.root_dir, 'usr', 'bin'])
+        }
+        rpm_bin = Path.which(
+            'rpm', custom_env=rpm_search_env, access_mode=os.X_OK
+        )
+        if not rpm_bin:
+            return False
+        return True
+
+    def has_consistent_database(self):
+        """
+        Check if image rpm database is consistent
+
+        :return: True or False
+
+        :rtype: bool
+        """
+        try:
+            Command.run(
+                ['chroot', self.root_dir, 'rpmdb', '--initdb']
+            )
+            return True
+        except Exception:
+            return False
+
+    def rebuild_database(self):
+        """
+        Rebuild image rpm database taking current macro setup into account
+        """
+        Command.run([
+            'chroot', self.root_dir, 'rpmdb', '--rebuilddb'
+        ])
+
+    def set_database_to_host_path(self):
+        """
+        Setup dbpath to point to the host rpm dbpath configuration
+        """
+        self.rpmdb_image.set_config_value(
+            '_dbpath', self.rpmdb_host.get_query('_dbpath')
+        )
+        self.rpmdb_image.write_config()
+
+    def set_database_to_image_path(self):
+        """
+        Setup dbpath to point to the image rpm dbpath configuration
+        """
+        self.rpmdb_image.wipe_config()
+        rpm_image_dbpath = self.rpmdb_image.expand_query('%_dbpath')
+        if rpm_image_dbpath != self.rpmdb_host.expand_query('%_dbpath'):
+            self.rpmdb_image.set_config_value(
+                '_dbpath_rebuild', self.rpmdb_image.get_query('_dbpath')
+            )
+            Path.wipe(
+                os.sep.join([self.root_dir, rpm_image_dbpath])
+            )
+            self.rpmdb_image.write_config()
+            self.rebuild_database()
+            self.rpmdb_image.wipe_config()
+
+    def dump_reload_database(self, version=45):
+        """
+        Dump and reload the rpm database to match the desired rpm db version
+        Supported target rpm database db versions are
+
+        * 45 (db45_load)
+        * 48 (db48_load)
+
+        :param str version: target rpm db version
+        """
+        db_load_for_version = {
+            45: 'db45_load',
+            48: 'db48_load'
```

Comment:
I know this is technically unchanged, but this does't work for non-SUSE distributions, because everyone else moved past db4. Both RHEL 7 and Fedora use db5 (as does Mageia). OpenMandriva is using db6, and so on.

Why do we do this anyway?

Reply:
Thanks for the discussion on the matrix chat. I deleted that reload code as it is not needed anymore